Blogpost: Motion Matching in O3DE, a Data-Driven Animation Technique

amzn-jillich opened the pull request:
Deploy preview here: https://deploy-preview-1591--o3deorg.netlify.app/blog/posts/blog-motionmatching/

### Submission Checklist:

* [x] **Descriptive active voice** - Do descriptive sentences have a clear *subject* and *action verb*?
* [x] **Answer the question at hand** - Does the documentation answer a *what*, *why*, *how*, or *where* type of question?
* [x] **Consistency** - Does the content consistently follow the [style guide](https://o3de.org/docs/contributing/to-docs/style-guide/)?
* [x] **Help the user** - Does the documentation show the user something *meaningful*?

micronAMZN commented:
We have shortcodes for embedding video and scaling images, `video` and `image-width` you can read more about them here: https://www.o3de.org/docs/contributing/to-docs/style-guide/media/.

I have a minor concern about using githubusercontent to host website content because I can't find any concrete information about their retention policy. Since this is a blog post, we could make an exception for file size. Using the first video as an example, I was able to reduce the size from about 85 MB to about 50 MB by removing the unused audio track it contains, and uniform scaling it to a width of 1080, which is the maximum size it scales on page on my system. It's over a minute long. If it's reasonable to demonstrate the feature in under 30 seconds I think I can reduce it to an acceptable file size. Similar feedback for the other videos. Basically, if it's possible to reduce the duration, I can process them and probably make them a reasonable size for the repo.

hultonha commented on content/blog/posts/blogpost-motion-matching.md:
````diff
@@ -0,0 +1,189 @@
+
+In case the new best matching frame candidate is close to the time in the animation that we are already playing, we don't do anything as we seem to be at the sweet spot in the motion database already.
+
+Pseudo-code:
+```
````
nit: I'd add a line break here

amzn-jillich replied:
Added

hultonha commented on content/blog/posts/blogpost-motion-matching.md:
```diff
@@ -0,0 +1,189 @@
+
+This makes motion matching an optimization problem where the frame with the minimum cost is the most preferred candidate to transition to.
+
+#### Searching next best matching frame
+The actual search happens in two phases, a broad phase to eliminate most of the candidates followed by a narrow phase to find the actual best candidate.
+
+##### 1. Broad-phase (KD-tree)
+A KD-tree is used to find the nearest neighbors (frames in the motion database) to the query vector (given input). The result is a set of pre-selected frames for the next best matching frame that is passed on to the narrow-phase. By adjusting the maximum tree depth or the minimum number of frames for the leaf nodes, the resulting number of frames can be adjusted. The bigger the set of frames the  ...
```
nit: It's a good idea to be consistent here and have line breaks after each heading (anything using `#`)

hultonha commented on content/blog/posts/blogpost-motion-matching.md:
````diff
@@ -0,0 +1,189 @@
+##### 2. Narrow-phase
+Inside the narrow-phase, we iterate through the returned set of frames from the KD-tree, and evaluate and compare their cost against each other. The frame with the minimal cost is the best match that we transition to.
+
+Pseudo-code:
+```
````
nit: line break here

hultonha reviewed (approved):
This looks good to me! 🙂

amzn-jillich commented on content/blog/posts/blogpost-motion-matching.md:
```diff
@@ -0,0 +1,189 @@
+---
+title: "Motion Matching in O3DE, a Data-Driven Animation Technique"
+date: 2022-04-28
```
Changed

amzn-jillich commented:
@micronAMZN I've changed the html to the markdown tags suggested in the style guide. They work great with the sizing! Also enabled auto play and looping for the videos.
I've also added the images to the repository + reduced the size of the videos significantly.

micronAMZN commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+
+{{< image-width src="/images/blog/motion-matching/script-canvas-graph.png" width="100%" alt="Script Canvas Graph for motion matching character control." >}}
+
+For an interactive game, we can use the gamepad's joystick position as a target to generate a desired future trajectory. By creating an exponential curve starting at the current character position going towards the character facing direction and then bend the curve towards the joystick's target position. This process is called trajectory prediction.
```
Had trouble parsing this one. The second sentence seems like a fragment as written.

```suggestion
For an interactive game, we can use the gamepad's joystick position as a target to generate a desired future trajectory by creating an exponential curve starting at the current character position going towards the character facing direction, and then bend the curve towards the joystick's target position. This process is called trajectory prediction.
```

micronAMZN commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+
+## Data Analysis & Debugging
+
+While motion matching gets rid of the manual process of setting up locomotion in an animation graph, precisely time transitions and makes sure the blends look good, it also takes away some control. Visualizing data and the state of the algorithm and knowing our data is key for debugging an animation glitch. We provide several debug visualizations: Features like a joint velocity or the trajectory  ...
```
```suggestion
While motion matching gets rid of the manual process of setting up locomotion in an animation graph, precisely times transitions, and makes sure the blends look good, it also takes away some control. Visualizing data and the state of the algorithm and knowing our data is key for debugging an animation glitch. We provide several debug visualizations: Features like a joint velocity or the trajectory can be visualized together with the query values as well as the feature costs along with some performance metrics. The feature costs over time can be used to see if a visual glitch corresponds with a high cost at that given point in time. A high cost indicates that the algorithm struggled to find a good matching pose for the query. And this again says that our motion database might lack a motion for the given scenario that can be fixed by retaking or adding an extra animation clip.
```

micronAMZN reviewed (approved):
This is fantastic. Thanks for addressing the videos.

Approving with some mostly style related suggestions.

amzn-jillich commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+
+Motion matching is a data-driven animation technique that is gaining popularity after games like 'The Last Of Us Part II', 'FIFA', and 'Half-Life: Alyx' started adopting it. The complexity of animation graphs has increased along with higher expectations, a greater variety of interactions between characters and their environments, and the need for more realistic animations. Motion matching is a wa ...
+
+We brought motion matching to O3DE's animation system EMotion FX as a blend tree node so that it can still be used and mixed with everything else available in the animation graph.
```
Changed

amzn-jillich commented:
Thanks for the suggestions @micronAMZN ! I've used the handy commit tool here directly but not we're getting DCO issues. Anyone knows what to do about this?

amzn-jillich commented:
DCO issues solved

aFinchy reviewed (approved):
Looks good, Approving this. Do not merge until May 12th 2022

hold/do-not-merge

micronAMZN commented:
> Thanks for the suggestions @micronAMZN ! I've used the handy commit tool here directly but not we're getting DCO issues. Anyone knows what to do about this?

For future reference., when you commit suggestions in the browser, you have to add the sign-off manually in the message box, and you might have to use the anonymized email found in your email settings.

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+full_img: "/images/blog/motion-matching/full_image.png"
+---
+
+Motion matching is a data-driven animation technique that is gaining popularity since games like 'The Last Of Us Part II', 'FIFA', and 'Half-Life: Alyx' have adotped it. The complexity of animation graphs has increased along with higher expectations, a greater variety of interactions between characters and their environments, and the need for more realistic animations. Motion matching is a way to ...
```
IMHO the second sentance lacks a little clarity " The complexity of animation graphs has increased along with higher expectations, a greater variety of interactions between characters and their environments, and the need for more realistic animations."

Maybe something along the lines of "[Users|Players|Gamers] have higher expectations for animation quality including more realistic animations and a greater variety of interactions between characters and their environments. This growth in expectation is matched by a growth in animation graph complexity which has made [balancing|generating|developing?] them more expensive.

nit: Is "locomotion" the right term in the last sentence? Would argue that not all anim graphs involve motion that moves a character from one place to another.

amzn-jillich replied:
Changed it to:

"Users have higher expectations for animation quality including more realistic animations and a greater variety of interactions between characters and their environments. This growth in expectation resulted in increasingly complex animation graphs that became harder to maintain and more expensive to create."

Does that sound ok?

lmbr-pip replied:
Sounds great

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+
+Motion matching is a data-driven animation technique that is gaining popularity since games like 'The Last Of Us Part II', 'FIFA', and 'Half-Life: Alyx' have adotped it. The complexity of animation graphs has increased along with higher expectations, a greater variety of interactions between characters and their environments, and the need for more realistic animations. Motion matching is a way to ...
+
+We brought motion matching to O3DE's animation system EMotion FX as a blend tree node so that it can still be used and mixed with everything else available in the animation graph.
```
```suggestion
We brought motion matching to O3DE's animation system [EMotion FX](https://docs.aws.amazon.com/lumberyard/latest/userguide/gems-system-gem-emotionfx-animation.html ) as a blend tree node so that it can still be used and mixed with everything else available in the animation graph.
```

amzn-jillich replied:
Used the O3DE documentation link here: https://www.o3de.org/docs/user-guide/visualization/animation/char-intro/
Or do you think it is better to use the LY link here?

amzn-jillich replied:
Added

lmbr-pip replied:
Either is good. Think its always good to provide context to system names, techniques, more uncommon terms etc for those are unfamiliar/new to the space.

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+
+## Motion Database
+
+To animate a character using motion matching, we need a large library of character animations such as walking slowly, running fast, narrow and wide turns, accelerating and decelerating, and many more. A large variety of movements are required so that there is a piece of animation for every situation. Kristjan Zadziuk and his team ([GDC 2016 Talk, Animation Bootcamp: Motion Matching](https://www.y ...
```
Nit: Maybe set some expectations about if we are providing a sample Motion Database ie "You can see a sample database of moves in.."? Or you will need to generate such a database, see our recommendations ... etc.

Not sure if this covers how users should store motions or set things up or where to look next.

amzn-jillich replied:
Added

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+
+## Motion Features
+
+A smooth animation is the result of continuity. Sudden changes in position or velocity will result in unrealistic motion. We want to synthesize an animation that is smooth and matches the intention of the user's interaction. If the user presses forward on the game-pad, it is not enough to transition to an animation sequence that results in a smooth animation while the character starts walking bac ...
```
I didn't quite the point of "If the user presses forward on the game-pad, it is not enough to transition to an animation sequence that results in a smooth animation while the character starts walking backward."

I think you are saying if the user presses forward, its not enough to transition to the forward move animation sequence if the character is currently walking backwards....

amzn-jillich replied:
Hm...no. I was trying to say that it is not enough to keep the animation smooth while ignoring the expected direction of movement is ignored. Changed it to that now.

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+
+In order to find the best possible transition and synthesize smooth animations, we need data that we call *features* to help us. A feature is a property extracted from the animation data that is used by the motion matching algorithm to find a good match for a transition. Examples of features are the position of the feet joints, the linear or angular velocity of the knee joints, or the trajectory  ...
+
+Their purpose is to describe the key characteristics of the pose and the animation and sometimes even enhance the actual key-frame data (the local-space joint transforms) by, for example, taking the time domain into account and calculating the velocity or acceleration, or a whole trajectory to describe where the given joint came from to reach the frame and the path it moves along in the near futu ...
```
nit: "Their .." is unclear. A feature, an environment sensation eetc.

amzn-jillich replied:
Changed to "A feature's purpose..."

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+| Matches joint positions | Matches joint velocities | Matches the trajectory history and future trajectory |
+| {{< image-width src="/images/blog/motion-matching/feature-position.png" width="100%" alt="Position Feature." >}} | {{< image-width src="/images/blog/motion-matching/feature-velocity.png" width="100%" alt="Velocity Feature." >}}  | {{< image-width src="/images/blog/motion-matching/feature-trajectory.png" width="100%" alt="Trajectory Feature." >}}  |
+
+For example, the interactive character in a game might walk in a different direction than in the pre-recorded animation which makes it impossible for us to work with world-space velocities. Features are extracted and stored relative to a given joint, in most cases the motion extraction or root joint, and thus are in model-space. This makes the search algorithm invariant to the character location  ...
```
```suggestion
For example, the interactive character in a game might walk in a different direction than in the pre-recorded animation, which makes it impossible for us to work with world-space velocities. Features are extracted and stored relative to a given joint, in most cases the motion extraction or root joint, and thus are in model-space. This makes the search algorithm invariant to the character location and orientation and the extracted features, e.g. a joint position or velocity translate and rotate along with the character.
```

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+
+There is no set of features that result in the highest quality animation in all scenarios. Different situations might require a different bias in the search and thus we want to be able to customize the algorithm. The feature schema is a set of features that define the criteria used in the motion matching algorithm and influences the runtime speed, memory usage, and the results of the synthesized  ...
+
+The schema defines which features are extracted from the motion database while the actual extracted data is stored in the feature matrix. Along with the feature type, settings like the joint to extract the data from, a debug visualization color, how the residual is calculated, or a custom feature is specified.
```
```suggestion
The schema defines which features are extracted from the motion database while the actual extracted data is stored in the feature matrix. Along with the feature type, settings like the joint to extract the data from, a debug visualization color, how the residual is calculated, or a custom feature can be specified.
```

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+
+The schema defines which features are extracted from the motion database while the actual extracted data is stored in the feature matrix. Along with the feature type, settings like the joint to extract the data from, a debug visualization color, how the residual is calculated, or a custom feature is specified.
+
+The more features selected by the user, the bigger the chances are that the searched and matched pose hits the expected result but the slower the algorithm will be and the more memory it will use. The key is to use crucial and independent elements that define a pose and its movement without being too strict on the wrong end. The root trajectory along with the left and right foot positions and vel ...
```
```suggestion
The more features selected by the user, the bigger the chances are that the searched and matched pose hits the expected result, but the slower the algorithm will be and the more memory it will use. The key is to use crucial and independent elements that define a pose and its movement without being too strict. The root trajectory along with the left and right foot positions and velocities have been proven to be a good start here.
```

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+
+Each row represents the features of a single frame of the motion database. As the data is commonly accessed on a per-frame basis, it is stored in row-major order for cache efficiency.
+
+> Memory usage: A motion capture database holding 1 hour of animation data together with a sample rate of 30 Hz to extract features, resulting in 108,000 frames, using the default feature schema having 59 features, will result in a feature matrix holding ~6.4 million values and use ~24.3 MB of memory.
```
```suggestion
> Memory usage: A motion capture database holding 1 hour of animation data together with a sample rate of 30 Hz to extract features will generate 108,000 frames. Using the default feature schema, comprising of 59 features, will result in a feature matrix holding ~6.4 million values and use ~24.3 MB of memory.
```

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+
+In the majority of the game ticks, the current motion gets advanced. A few times per second, the actual motion matching search is triggered to prevent drifting too far away from the expected user input (as we would just play the recorded animation otherwise).
+
+When a search for a better next matching frame is triggered, the current pose, including its joint velocities, gets evaluated. This pose (which we'll call input or query pose) is used to fill the query vector. The query vector contains feature values and is compared against other frames in the feature matrix. The query vector has the same size as there are columns in the feature matrix and is sim ...
```
Am not quite to sure what you mean by "... and is similar to any other row but represents the input query."

amzn-jillich replied:
Changed to: "The query vector has the same size as there are columns in the feature matrix and is similar to any other row which stores the features for a given frame in the motion database."

Does that make it more clear?

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+
+## Character trajectories
+
+Let's talk about trajectories. A trajectory is a useful way to encode where the character came from and its intention of where and how it wants to move in the future. We store a number of samples containing character positions as well as facing directions for a given time window. For example, 4 samples into the past across 0.7 seconds and 6 samples for the desired future trajectory across 1.2 sec ...
```
```suggestion
Let's talk about trajectories. A trajectory is a useful way to encode where the character came from and its intention of where and how it wants to move in the future. We store a number of samples containing character positions, as well as facing directions, for a given time window. For example, 4 samples into the past across 0.7 seconds and 6 samples for the desired future trajectory across 1.2 seconds. The Euclidean distances between the samples tell us if the character is accelerating or coming to a stop. The sample locations indicate if the character is moving forward or doing a left turn, and the facing direction gives us more information about if the character is strafing or moving backwards. Recording the past trajectory is easy as we know the current character transform, but how do we decide the future trajectory to generate, and why do we even need that?
```

lmbr-pip replied:
minor: Its odd to end in a questions esp when the content is immediately followed by an image.

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+
+## Data Analysis & Debugging
+
+While motion matching gets rid of the manual process of setting up locomotion in an animation graph, precisely time transitions and makes sure the blends look good, it also takes away some control. Visualizing data and the state of the algorithm and knowing our data is key for debugging an animation glitch. We provide several debug visualizations: Features like a joint velocity or the trajectory  ...
```
```suggestion
While motion matching gets rid of the manual process of setting up locomotion in an animation graph, precisely timing transitions and making sure the blends look good, it also takes away some control. Visualizing data and the state of the algorithm and knowing our data is key for debugging an animation glitchs. 

We provide several debug visualizations: Features like a joint velocity or the trajectory can be visualized together with the query values, as well as the feature costs along with some performance metrics. The feature costs over time can be used to see if a visual glitch corresponds with a high cost at that given point in time. A high cost indicates that the algorithm struggled to find a good matching pose for the query.  The problem matching poses may indicate that our motion database lacks a motion for the given scenario that can be fixed by retaking or adding an extra animation clip.
```

amzn-jillich replied:
Changed

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+
+While motion matching gets rid of the manual process of setting up locomotion in an animation graph, precisely time transitions and makes sure the blends look good, it also takes away some control. Visualizing data and the state of the algorithm and knowing our data is key for debugging an animation glitch. We provide several debug visualizations: Features like a joint velocity or the trajectory  ...
+
+We also provide data analysis tools to get some insight about what is missing or where we have duplicated animation data. In the middle image below you can see histograms per feature component showing their value distributions across the motion database. They can provide interesting insights, like for example, if the motion database is holding more moving forward animations than it has strafing o ...
```
```suggestion
We also provide data analysis tools to get some insight about what is missing or where we have duplicated animation data. In the middle image below you can see histograms per feature component showing their value distributions across the motion database. They can provide interesting insights, for example, if the motion database is holding more moving forward animations than it has strafing or backward moving animations, or how many fast vs slow turning animations are in the database. This information can be used to see if there is still a need to record some animations or if some types of animation are over-represented and will lead to ambiguity and decrease the quality of the resulting synthesized animation.
```

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+
+We also provide data analysis tools to get some insight about what is missing or where we have duplicated animation data. In the middle image below you can see histograms per feature component showing their value distributions across the motion database. They can provide interesting insights, like for example, if the motion database is holding more moving forward animations than it has strafing o ...
+
+The image on the right shows a scatter plot of the features per frame in the motion database. Our multi-dimensional feature data is projected down to two dimensions using principal component analysis. The density of the clusters and the distribution of the samples overall indicate how hard it is for the search algorithm to find a good matching frame candidate. It shows what areas are well covered ...
```
Considering adding wikipedia or other link to PCA. It may be a new concept esp for traditional animators

amzn-jillich replied:
Added

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,179 @@
+{{< image-width src="/images/blog/motion-matching/data-analysis.png" width="100%" alt="Debug visualization and data analysis." >}}
+
+## Future Outlook
+Research in the field of animation has advanced dramatically over the last several years. Traditional animation techniques are melding together, being replaced or enhanced by machine learning. Characters are becoming more aware of their environments, automatically reacting to stimuli and are actually now physically present in the world. Control is moving from simple directional input to high-leve ...
```
Would argue characters are not "physically present in the world" more that they are more physically centered in the world.

lmbr-pip reviewed (commented):
Looks really good.

There were a couple of places I stumbled over the text, but some maybe because I just don't have the technical background here.

I added some suggestions which you should feel free to ignore, but they generally highlighted places where reading out loud highlighted some awkwardness (at least for me).

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,181 @@
+
+## Motion Features
+
+A smooth animation is the result of continuity. Sudden changes in position or velocity will result in unrealistic motion. We want to synthesize an animation that is smooth and matches the intention of the user's interaction. If the user presses forward on the game-pad, it is not enough to transition to an animation sequence that results in a smooth animation while the expected direction of moveme ...
```
nit: We use "smooth" a few times in this paragraph and below. May want to consider alternatives just to avoid repetition.

amzn-jillich replied:
Changed

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,181 @@
+
+In order to find the best possible transition and synthesize smooth animations, we need data that we call *features* to help us. A feature is a property extracted from the animation data that is used by the motion matching algorithm to find a good match for a transition. Examples of features are the position of the feet joints, the linear or angular velocity of the knee joints, or the trajectory  ...
+
+A feature's purpose is to describe the key characteristics of the pose and the animation and sometimes even enhance the actual key-frame data (the local-space joint transforms) by, for example, taking the time domain into account and calculating the velocity or acceleration, or a whole trajectory to describe where the given joint came from to reach the frame and the path it moves along in the nea ...
```
Minor: This is a fairly long sentence. Consider breaking up. My poor attempt below:

```suggestion
A feature's purpose is to describe the key characteristics of the pose and the animation. Sometimes features even enhance the actual key-frame data (the local-space joint transforms) by, for example, taking the time domain into account and calculating the velocity or acceleration. They can encode a whole trajectory to describe where the given joint came from to reach the frame and the path it moves along in the near future.
```

amzn-jillich replied:
Changed

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,181 @@
+
+While motion matching gets rid of the manual process of setting up locomotion in an animation graph, precisely timing transitions and making sure the blends look good, it also takes away some control. Visualizing data and the state of the algorithm and knowing our data is key for debugging an animation glitch.
+
+We provide several debug visualizations: Features like a joint velocity or the trajectory can be visualized together with the query values as well as the feature costs along with some performance metrics. The feature costs over time can be used to see if a visual glitch corresponds with a high cost at that given point in time. A high cost indicates that the algorithm struggled to find a good matc ...
```
nit: theres not really a list here, so remove use of ':' and maybe focus just on key debug visualizations. "several debug visualizations" is a little passive. Maybe "O3DE provides a number of motion matching visualizations to aid troubleshooting. "You can display features, for example joint velocities or the trajectory...:"

amzn-jillich replied:
Changed

lmbr-pip commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,181 @@
+
+We also provide data analysis tools to get some insight about what is missing or where we have duplicated animation data. In the middle image below you can see histograms per feature component showing their value distributions across the motion database. They can provide interesting insights, for example, if the motion database is holding more moving forward animations than it has strafing or bac ...
+
+The image on the right shows a scatter plot of the features per frame in the motion database. Our multi-dimensional feature data is projected down to two dimensions using [principal component analysis](https://en.wikipedia.org/wiki/Principal_component_analysis). The density of the clusters and the distribution of the samples overall indicate how hard it is for the search algorithm to find a good  ...
```
nit: Avoid starting with "It" in the final sentence as it forces me to think about what we are discussing. Maybe "The plot data .."?

amzn-jillich replied:
Changed to "graph"

lmbr-pip reviewed (commented):
Overall great.

Added a little more feedback but mostly this blog is in great shape and could ship as is.

amzn-jillich commented on content/blog/posts/o3de-motion-matching-2022-04.md:
```diff
@@ -0,0 +1,181 @@
+
+## Motion Database
+
+To animate a character using motion matching, we need a large library of character animations such as walking slowly, running fast, narrow and wide turns, accelerating and decelerating, and many more. A large variety of movements are required so that there is a piece of animation for every situation. Kristjan Zadziuk and his team ([GDC 2016 Talk, Animation Bootcamp: Motion Matching](https://www.y ...
```
Changed

chanmosq commented:
Hold from merging until 5/12 10:30am PST

amzn-jillich commented:
![image](https://user-images.githubusercontent.com/43751992/179942626-35daa9c5-9420-4003-831c-8aa60f4531df.png)
